test/qa/test_dns.py

```diff
@@ -84,23 +153,183 @@ def test_custom_dns_connect(tech, proto, obfuscated, nameserver):
+    nameserver = nameserver.split(" ")
+
+    lib.set_technology_and_protocol(tech, proto, obfuscated)
+
+    with lib.Defer(sh.nordvpn.disconnect):
+        sh.nordvpn.connect()
+        sh.nordvpn.set.dns(nameserver)
+
+        assert settings.dns_visible_in_settings(nameserver)
+        assert dns.is_set_for(nameserver)
+
+    assert dns.is_unset()
+
+
+@pytest.mark.parametrize("nameserver", dns.DNS_CASES_CUSTOM)
+@pytest.mark.parametrize("tech,proto,obfuscated", lib.TECHNOLOGIES)
+@pytest.mark.flaky(reruns=2, reruns_delay=90)
+@timeout_decorator.timeout(40)
+def test_set_custom_dns_off_connected(tech, proto, obfuscated, nameserver):
+    nameserver = nameserver.split(" ")
+
+    lib.set_technology_and_protocol(tech, proto, obfuscated)
+
+    with lib.Defer(sh.nordvpn.disconnect):
+        sh.nordvpn.connect()
+
+        sh.nordvpn.set.dns(nameserver)
+        assert settings.dns_visible_in_settings(nameserver)
+        assert dns.is_set_for(nameserver)
+
+        sh.nordvpn.set.dns("off")
         assert settings.dns_visible_in_settings("disabled")
         assert dns.is_set_for(dns.DNS_NORD)
 
-        lib.set_threat_protection_lite("on")
-        assert settings.get_is_tpl_enabled()
+    assert dns.is_unset()
+
+
+@pytest.mark.parametrize("nameserver,expected_error", dns.DNS_CASES_ERROR)
+@pytest.mark.parametrize("tech,proto,obfuscated", lib.TECHNOLOGIES)
+@pytest.mark.flaky(reruns=2, reruns_delay=90)
+@timeout_decorator.timeout(40)
+def test_set_custom_dns_errors_disconnected(tech, proto, obfuscated, nameserver, expected_error):
+    lib.set_technology_and_protocol(tech, proto, obfuscated)
+
+    with pytest.raises(sh.ErrorReturnCode_1) as ex:
+        sh.nordvpn.set.dns(nameserver)
+
+    assert expected_error in str(ex)
+    assert settings.dns_visible_in_settings("disabled")
+    assert dns.is_unset()
+
+
+@pytest.mark.parametrize("nameserver,expected_error", dns.DNS_CASES_ERROR)
+@pytest.mark.parametrize("tech,proto,obfuscated", lib.TECHNOLOGIES)
+@pytest.mark.flaky(reruns=2, reruns_delay=90)
+@timeout_decorator.timeout(40)
+def test_set_custom_dns_errors_connected(tech, proto, obfuscated, nameserver, expected_error):
+    lib.set_technology_and_protocol(tech, proto, obfuscated)
+
+    with lib.Defer(sh.nordvpn.disconnect):
+        sh.nordvpn.connect()
+
+        with pytest.raises(sh.ErrorReturnCode_1) as ex:
+            sh.nordvpn.set.dns(nameserver)
+
+        assert expected_error in str(ex)
         assert settings.dns_visible_in_settings("disabled")
-        assert dns.is_set_for(dns.DNS_TPL)
+        assert dns.is_set_for(dns.DNS_NORD)
+
+    assert dns.is_unset()
+
+
+@pytest.mark.parametrize("nameserver", dns.DNS_CASES_CUSTOM)
+@pytest.mark.parametrize("tech,proto,obfuscated", lib.TECHNOLOGIES)
+@pytest.mark.flaky(reruns=2, reruns_delay=90)
+@timeout_decorator.timeout(40)
+def test_set_custom_dns_already_set_disconnected(tech, proto, obfuscated, nameserver):
```

Comment:
This is a very minor one, but the name of this test confused me, had to read the body to understand what is being tested. So now I know that this test tests what happens when setting custom dns while it is already set and while we are connected to VPN. So my suggestion for the name fix would be "test_set_custom_dns_while_already_set_and_connected". Now I know that this name is long, but as far as I know in test automation industry long names are generally OK and even preferable if it is possible to understand exactly what is being tested just from the name. Comment is applicable to other test names as well.